core/domain/learner_progress_services.py

```diff
@@ -616,3 +650,114 @@ def get_collection_summary_dicts(collection_summaries):
+            particular learner.
+        dict. The number of entities that are no longer present.
+        list(str). The titles of the collections to which new explorations have
+            been added.
+
+    """
+    incomplete_exploration_ids = get_all_incomplete_exp_ids(user_id)
+    incomplete_collection_ids = get_all_incomplete_collection_ids(user_id)
+    completed_collection_ids = get_all_completed_collection_ids(user_id)
+    completed_exploration_ids = get_all_completed_exp_ids(user_id)
+
+    incomplete_exploration_keys = (
+        datastore_services.get_keys_from_ids_and_model(
+            incomplete_exploration_ids, 'ExpSummaryModel'))
+    incomplete_collection_keys = (
+        datastore_services.get_keys_from_ids_and_model(
+            incomplete_collection_ids, 'CollectionSummaryModel'))
+    completed_exploration_keys = (
+        datastore_services.get_keys_from_ids_and_model(
+            completed_exploration_ids, 'ExpSummaryModel'))
+    completed_collection_keys = (
+        datastore_services.get_keys_from_ids_and_model(
+            completed_collection_ids, 'CollectionSummaryModel'))
+
+    activity_models = datastore_services.fetch_multiple_entities_by_keys(
+        incomplete_exploration_keys + incomplete_collection_keys +
+        completed_exploration_keys + completed_collection_keys)
+
+    incomplete_exploration_models = (
+        activity_models[:len(incomplete_exploration_ids)])
+    incomplete_collection_models = (
+        activity_models[
+            len(incomplete_exploration_ids):len(incomplete_exploration_ids) +
+            len(incomplete_collection_ids)])
+    completed_exploration_models = (
+        activity_models[
+            len(incomplete_exploration_ids) +
+            len(incomplete_collection_ids):len(incomplete_exploration_ids) +
+            len(incomplete_collection_ids) + len(completed_exploration_ids)])
+    completed_collection_models = (
+        activity_models[
+            len(incomplete_exploration_ids) +
+            len(incomplete_collection_ids) + len(completed_exploration_ids):])
+
+    incomplete_exp_summaries = (
+        [exp_services.get_exploration_summary_from_model(model)
+         if model else None for model in incomplete_exploration_models])
+    incomplete_collection_summaries = (
+        [collection_services.get_collection_summary_from_model(model)
+         if model else None for model in incomplete_collection_models])
+    completed_exp_summaries = (
+        [exp_services.get_exploration_summary_from_model(model)
+         if model else None for model in completed_exploration_models])
+    completed_collection_summaries = (
+        [collection_services.get_collection_summary_from_model(model)
+         if model else None for model in completed_collection_models])
+
+    filtered_incomplete_exp_summaries, num_deleted_incomplete_exps = (
+        get_filtered_incomplete_exp_summaries(
```

Comment:
This get() call does removals. GET calls should not have side-effects.

Instead, do:

    filtered_incomplete_exp_summaries, nonexistent_incomplete_exp_ids = get_filtered_incomplete_exp_summaries(...)
    remove_exp_ids_from_incomplete_list(nonexistent_incomplete_exp_ids)

Ditto below.